components/support/nimbus-fml/src/parser.rs

```diff
@@ -385,6 +614,91 @@ mod unit_tests {
+        let path_buf = Path::new(&path);
+        let parser = Parser::new(&Default::default(), path_buf)?;
+        let ir = parser.get_intermediate_representation()?;
+        let feature_def = ir.feature_defs.first().unwrap();
+        let positive_button = feature_def
+            .props
+            .iter()
+            .find(|x| x.name == "positive")
+            .unwrap();
+        // We validate that the no-channel feature level default got merged back
+        assert_eq!(
+            positive_button
+                .default
+                .get("alt-text")
+                .unwrap()
+                .as_str()
+                .unwrap(),
+            "Go Ahead!"
+        );
+        // We validate that the orignal field level default don't get lost if no
+        // feature level default with the same name exists
+        assert_eq!(
+            positive_button
+                .default
+                .get("label")
+                .unwrap()
+                .as_str()
+                .unwrap(),
+            "Ok then"
+        );
+        // We validate that feature level default overwrite field level defaults if one exists
+        // in the field level, it's blue, but on the feature level it's green
+        assert_eq!(
+            positive_button
+                .default
+                .get("color")
+                .unwrap()
+                .as_str()
+                .unwrap(),
+            "green"
+        );
+        // We now re-run this, but merge back the nightly channel instead
+        let config: Config = Config {
+            channel: Some("nightly".into()),
+            ..Default::default()
+        };
+        let parser = Parser::new(&config, path_buf)?;
+        let ir = parser.get_intermediate_representation()?;
+        let feature_def = ir.feature_defs.first().unwrap();
+        let positive_button = feature_def
+            .props
+            .iter()
+            .find(|x| x.name == "positive")
+            .unwrap();
+        // We validate that feature level default overwrite field level defaults if one exists
+        // in the field level, it's blue, but on the feature level it's bright-red
+        // note that it's bright-red because we merged back the `nightly`
+        // channel, instead of the `release` channel that merges back
+        // by default
+        assert_eq!(
+            positive_button
+                .default
+                .get("color")
+                .unwrap()
+                .as_str()
+                .unwrap(),
+            "bright-red"
+        );
+        // We againt validate that regardless
+        // of the channel, the no-channel feature level default got merged back
+        assert_eq!(
+            positive_button
+                .default
+                .get("alt-text")
+                .unwrap()
+                .as_str()
+                .unwrap(),
+            "Go Ahead!"
+        );
+        Ok(())
+    }
```

Comment:
The alogrithm you've sketched out has a number of parts/functions

For each of those parts you should write a test in which you try to explore the limits of what the function is meant to do.

When you come to put the well-tested parts together, you'll have more confidence in the pieces of the algorithm, and you'll be happy with a smoke test to test they fit together.